src/analysis/ml/ml.py

```diff
@@ -0,0 +1,137 @@
+from scipy.cluster.hierarchy import dendrogram
+import numpy as np
+
+def summarize_networks(file_paths: Iterable[Path]) -> pd.DataFrame:
+
+    # creating a tuple that contains the algorithm column name and edge pairs
+    edge_tuples = []
+    
+    for file in file_paths:
+        
+        try:
+           
+           # collecting and sorting the edge pairs per algortihm
+            with open(file,'r') as f:
+                # under the assumption that it is single chars (maybe)
+                # split on space char 
+                # take the first and 0th elements and sort those 
+                lines = [line[:3] for line in f.readlines()]
+            
+            line = []
+            for char in lines:
+                newChar = char.replace(' ','')
+                line.append(newChar)
+                
+            # to deal with edges are the same but not ordered the same
+            e = [''.join(sorted(ele)) for ele in line]
+
+            
+            # getting the algorithm name
+            p = PurePath(file)
+            p.parts[-2]
+            edge_tuples.append((p.parts[-2], e))
+
+  
+        except FileNotFoundError:
+            print(file, ' DOES NOT EXIST') # should hopefully not hit
+            continue
+        
+    # the dataframes per algorithm
+    edge_dataframes = []
+    
+    # the data frame is set up per algorithm and a 1 is set for the edge pair that exists in the algorithm
+    for tuple in edge_tuples:
+
+        dataframe = pd.DataFrame(
+            {
+                str(tuple[0]): 1,
+            }, index= tuple[1]
+        )
+        edge_dataframes.append(dataframe)
+
+    # concating all the data frames together (0 is set for all the edge pairs that don't exist per algorithm)
+    concated_df = pd.concat(edge_dataframes, axis= 1, join = 'outer')
+    concated_df = concated_df.fillna(0) 
+    concated_df = concated_df.astype('int64')
+   
+    return concated_df
+    
+def pca(dataframe: pd.DataFrame, output_png: str, output_file: str):
+
+    df = dataframe.reset_index(drop=True)
+    df = df.transpose() #based on the algortihms rather than the edges 
+    X = df.values
+
+    scaler = StandardScaler()
+    scaler.fit(X) # calc mean and standard deviation 
+    X_scaled = scaler.transform(X)
+
+    # chosing the PCA
+    pca_2 = PCA(n_components=2)
+    pca_2.fit(X_scaled)
+    X_pca_2 = pca_2.transform(X_scaled)
+    variance = pca_2.explained_variance_ratio_ *100
+
+    # making the plot
+    plt.figure(figsize = (10,7))
+    # sns.set(font_scale = 1.5)
+    sns.scatterplot(x = X_pca_2[:,0], y = X_pca_2[:,1], s = 70)
+    plt.title("PCA")
+    plt.xlabel(f"PC1 ({variance[0]:.1f}% variance)")
+    plt.ylabel(f"PC2 ({variance[1]:.1f}% variance)")
+    
+    # saving the pca plot
+    plt.savefig(output_png)
+
+    # saving the principal components
+    with open(output_file, "w") as f: 
+        f.write(str(variance))
+   
+# site sckit learn 
```

Comment:
Still want to do this

Reply:
Because we are going to use code directly, I'll make a commit that adds appropriate acknowledgement here and in the main readme.